source/include/actbl2.h

```diff
@@ -2265,6 +2265,215 @@ typedef struct nfit_device_handle
     (((handle) & ACPI_NFIT_NODE_ID_MASK)        >> ACPI_NFIT_NODE_ID_OFFSET)
 
 
+/*******************************************************************************
+ *
+ * NHLT - Non HDAudio Link Table
+ *        Version 1
+ *
+ ******************************************************************************/
+
+typedef struct acpi_table_nhlt
+{
+    ACPI_TABLE_HEADER              Header;          /* Common ACPI table header */
+    UINT8                          EndpointsCount;
+    /*
+     * ACPI_NHLT_ENDPOINT          Endpoints[];
+     * ACPI_NHLT_CONFIG            OEDConfig;
+     */
+
+} ACPI_TABLE_NHLT;
+
+typedef struct acpi_nhlt_endpoint
+{
+    UINT32                         Length;
+    UINT8                          LinkType;
+    UINT8                          InstanceId;
+    UINT16                         VendorId;
+    UINT16                         DeviceId;
+    UINT16                         RevisionId;
+    UINT32                         SubsystemId;
+    UINT8                          DeviceType;
+    UINT8                          Direction;
+    UINT8                          VirtualBusId;
+    /*
+     * ACPI_NHLT_CONFIG            DeviceConfig;
+     * ACPI_NHLT_FORMATS_CONFIG    FormatsConfig;
+     * ACPI_NHLT_DEVICES_INFO      DevicesInfo;
+     */
+
+} ACPI_NHLT_ENDPOINT;
+
+/* Values for LinkType field above */
+
+#define ACPI_NHLT_LINKTYPE_HDA               0
+#define ACPI_NHLT_LINKTYPE_DSP               1
+#define ACPI_NHLT_LINKTYPE_PDM               2
+#define ACPI_NHLT_LINKTYPE_SSP               3
+#define ACPI_NHLT_LINKTYPE_SLIMBUS           4
+#define ACPI_NHLT_LINKTYPE_SDW               5
+#define ACPI_NHLT_LINKTYPE_UAOL              6
+
+/* Values for DeviceId field above */
+
+#define ACPI_NHLT_DEVICEID_DMIC              0xAE20
+#define ACPI_NHLT_DEVICEID_BT                0xAE30
+#define ACPI_NHLT_DEVICEID_I2S               0xAE34
+
+/* Values for DeviceType field above */
+
+/* Device types unique to endpoint of LinkType=PDM */
+#define ACPI_NHLT_DEVICETYPE_PDM             0
+#define ACPI_NHLT_DEVICETYPE_PDM_SKL         1
+/* Device types unique to endpoint of LinkType=SSP */
+#define ACPI_NHLT_DEVICETYPE_BT              0
+#define ACPI_NHLT_DEVICETYPE_FM              1
+#define ACPI_NHLT_DEVICETYPE_MODEM           2
+#define ACPI_NHLT_DEVICETYPE_CODEC           4
+
+/* Values for Direction field above */
+
+#define ACPI_NHLT_DIR_RENDER                 0
+#define ACPI_NHLT_DIR_CAPTURE                1
+
+typedef struct acpi_nhlt_config
+{
+    UINT32                         CapabilitiesSize;
+    UINT8                          Capabilities[];
```

Comment:
We need a bound on Capabilities[] flexible array since the struct object is in itself made into another flexible array from inside another struct! So I recommend something like Capabilities[1024] here which should ensure that the 1024 is logically a big enough number for any practical applications/purposes. Thanks!

Reply:
This feels simply wrong. I believe you have to have it as flexible array. Can't you use `ACPI_FLEX_ARRAY()`?

Reply:
It would look more organized for sure, but then we are opening a can of worms since there are dozens of such usages all over in ACPICA and maybe one day we can clean them up! AFAIK ACPI_FLEX_ARRAY() was added fairly recently because GCC 12(+) or Clang would not allow (without a warning and thus breaking our build) a flexible array as the sole element of a struct/union that is nested inside another struct/union. As a side effect, we have to disable a specific MSVC warning C4201 on nameless structs/unions usage too. This merged PR added the ACPI_FLEX_ARRAY() feature:
https://github.com/acpica/acpica/pull/856

However, in this scenario, even using ACPI_FLEX_ARRAY() does not work because under the hood it is still expanded to [] or [0] depending on the environment and MSVC simply does not allow any type of usage of nested flexible arrays which results in that error C2233. I recommend Cezary updates the bound on Capabilities[] both in code and documentation and maybe uses a macro instead of a magic number like I used to get past the hurdle, but I am still afraid he won't be able to use any kind of flexible and nested arrays! It feels like a hard job sometimes for us balancing both MSVC and GCC/LLVM/Clang compilers and keeping all the environments stable and warning/error-free!

Reply:
I didn't drink my coffee and replied in general thread rather than this conversion. Sorry for the mess.

Reply:
> It would look more organized for sure, but then we are opening a can of worms since there are dozens of such usages all over in ACPICA and maybe one day we can clean them up! AFAIK ACPI_FLEX_ARRAY() was added fairly recently because GCC 12(+) or Clang would not allow (without a warning and thus breaking our build) a flexible array as the sole element of a struct/union that is nested inside another struct/union. As a side effect, we have to disable a specific MSVC warning C4201 on nameless structs/unions usage too. This merged PR added the ACPI_FLEX_ARRAY() feature: #856
> 
> However, in this scenario, even using ACPI_FLEX_ARRAY() does not work because under the hood it is still expanded to [] or [0] depending on the environment and MSVC simply does not allow any type of usage of nested flexible arrays which results in that error C2233. I recommend Cezary updates the bound on Capabilities[] both in code and documentation and maybe uses a macro instead of a magic number like I used to get past the hurdle, but I am still afraid he won't be able to use any kind of flexible and nested arrays! It feels like a hard job sometimes for us balancing both MSVC and GCC/LLVM/Clang compilers and keeping all the environments stable and warning/error-free!

I would rather say that MSVC is simply broken. And maybe exclude compilation of NHLT from the ACPICA on Windows (bugs or misfeatures in the compilers are not so rare).

Reply:
I might have been able to fix it using Capabilities[1] on MSVC after updating the toolset on MSVC which includes the C/C++ compiler. The most recent commit is addressed to that effect and I am hoping that with GCC version >12 it won't throw a warning thus breaking the build since I found similar usages (1 in the same file) and some in other parts of ACPICA (I can't test GCC>12 on my old NUC with Ubuntu).